// xrpl-accountlib/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.rawSigning = exports.utils = exports.sign = exports.derive = exports.generate = exports.XRPL_Account = void 0;
/* Methods  ==================================================================== */
var generate = __importStar(require("./generate"));
exports.generate = generate;
var derive = __importStar(require("./derive"));
exports.derive = derive;
var utils = __importStar(require("./utils"));
exports.utils = utils;
var sign_1 = __importDefault(require("./sign"));
exports.sign = sign_1.default;
var rawSigning = __importStar(require("./rawSigning"));
exports.rawSigning = rawSigning;
/* Types ==================================================================== */
var Account_1 = __importDefault(require("./schema/Account"));
exports.XRPL_Account = Account_1.default;
